tools/sql_extraction/src/main/kotlin/Environment.kt

```diff
@@ -0,0 +1,96 @@
+package com.google.cloud.sqlecosystem.sqlextraction
+
+import com.google.cloud.sqlecosystem.sqlextraction.output.QueryFragment
+
+/**
+ * A mapping from variable name to all possible query fragment values
+ */
+class Environment {
+    private var parentScope: Environment?
+    private var reachingDefs: HashMap<String, QueryFragment?>
+
+    constructor() {
+        this.parentScope = null
+        this.reachingDefs = HashMap()
+    }
+
+    private constructor(copy: Environment) {
+        this.parentScope = copy.parentScope
+        this.reachingDefs = copy.reachingDefs
+    }
+
+    /**
+     * Declare [varName] as a newly declared variable. Its initial reaching query is null.
+     * [setVariable] only runs successfully for declared variables.
+     */
+    fun declareVariable(varName: String) {
+        reachingDefs[varName] = null
+    }
+
+    /**
+     * Checks whether a variable of name [varName] was declared in this current scope.
+     *
+     * @return true if variable was declared in this scope
+     */
+    fun hasVariableInScope(varName: String): Boolean {
+        return varName in reachingDefs
+    }
+
+    /**
+     * Gets all possible queries for the variable [varName].
+     * Variable can be declared in any reachable scope.
+     *
+     * @throws[NullPointerException] if variable does not exist
+     */
+    fun getVariable(varName: String): QueryFragment? {
+        return if (hasVariableInScope(varName)) {
+            reachingDefs[varName]
+        } else {
+            parentScope!!.getVariable(varName)
+        }
+    }
+
+    /**
+     * Gets all possible queries for the variable [varName].
+     * Variable can be declared in any reachable scope.
+     * [default] is returned if variable doesn't exist in any reachable scope.
+     */
+    fun getVariableOrDefault(varName: String, default: QueryFragment? = null): QueryFragment? {
+        return when {
+            hasVariableInScope(varName) -> reachingDefs[varName]
+            parentScope != null -> parentScope!!.getVariableOrDefault(varName)
```

Comment:
Seems `parentScope!!.getVariableOrDefault(varName)` should be `parentScope!!.getVariableOrDefault(varName, default)`?

Reply:
Oh oops, thanks for pointing that out!